spec.html

```diff
@@ -35823,13 +35863,25 @@ <h1>
         <emu-grammar>Atom :: CharacterClass</emu-grammar>
         <emu-alg>
           1. Let _cc_ be CompileCharacterClass of |CharacterClass|.
-          1. If _UnicodeSets_ is *false* or if every element of _cc_.[[CharSet]] consists of a single character, return ! CharacterSetMatcher(_cc_.[[CharSet]], _cc_.[[Invert]], _direction_).
+          1. Let _cs_ be _cc_.[[CharSet]].
+          1. If _UnicodeSets_ is *false* or if every element of _cs_ consists of a single character, return ! CharacterSetMatcher(_cs_, _cc_.[[Invert]], _direction_).
           1. Assert: _cc_.[[Invert]] is *false*.
-          1. Let _D_ be the Disjunction containing:
-            1. an Alternative for each string in _cc_ whose length is greater than 1, with longer strings before shorter strings.
-            1. an Alternative for the CharSet containing all and only the single characters in _cc_ (that is, all and only the strings of length 1).
-            1. an Alternative for the empty String, if it is in _cc_.
-          1. Return CompileSubpattern of _D_ with argument _direction_.
+          1. Let _lm_ be an empty list of Matchers.
+          1. For each string _s_ in _cs_ whose length is greater than 1, iterating in descending order of string length, do
+            1. Let _cs2_ be a one-element CharSet containing the last code point of _s_.
+            1. Let _m2_ be the result of ! CharacterSetMatcher(_cs2_, *false*, _direction_).
+            1. For each code point _c1_ in _s_, iterating backwards from its second-to-last code point, do
+              1. Let _cs1_ be a one-element CharSet containing _c1_.
+              1. Let _m1_ be the result of ! CharacterSetMatcher(_cs1_, *false*, _direction_).
+              1. Let _m2_ be ! MatchSequence(_m1_, _m2_).
+            1. Append _m2_ to _lm_.
+          1. Let _singles_ be the CharSet that contains every element of _cc_.[[CharSet]] which consists of a single character.
```

Comment:
Nit: there's no reason to separate out this case. If you iterate over all strings of length greater than 0, rather than strings of length greater than 1, the previous loop covers these as well without modification.

Reply:
```suggestion
          1. Let _lm_ be an empty list of Matchers.
          1. For each string _s_ in _cs_ whose length is greater than 0, iterating in descending order of string length, do
            1. Let _cs2_ be a one-element CharSet containing the last code point of _s_.
            1. Let _m2_ be the result of ! CharacterSetMatcher(_cs2_, *false*, _direction_).
            1. For each code point _c1_ in _s_, iterating backwards from its second-to-last code point, do
              1. Let _cs1_ be a one-element CharSet containing _c1_.
              1. Let _m1_ be the result of ! CharacterSetMatcher(_cs1_, *false*, _direction_).
              1. Let _m2_ be ! MatchSequence(_m1_, _m2_).
            1. Append _m2_ to _lm_.
```

Reply:
I prefer to handle the singles separately.
- I think it's less confusing this way.
- Any real implementation will have a special path for matching the single code points, so this makes the spec look a bit more real.
- It's only two extra lines, and without them the inner loop above gets more complicated, as you discussed.

Note that in Mathias' version the lines after "Append _m2_ to _lm_." are indented too far.